File: Swift/Miscellaneous/3417-zigzag-grid-traversal-with-skip.swift
class Solution {
    func zigzagTraversal(_ grid: [[Int]]) -> [Int] {
        if (grid.isEmpty) {
            return []
        }
        
        if (grid.count == 1 && grid[0].count == 1) {
            return [grid[0][0]]
        }
        
        var result: [Int] = []
        
        for r in 0...grid.count-1 {

            if (r % 2 == 0) {
                var sI = 0
                var eI = grid[0].count-1
                while (sI <= eI) {
                    result.append(grid[r][sI])
                    sI += 2
                }
                
            } else {
                var sI = (grid[0].count % 2 == 0) ? grid[0].count-1 : grid[0].count-2
                
                while (sI >= 0) {
                    result.append(grid[r][sI])
                    sI -= 2
                }
            }
        }
        
        return result
    }
}

File: Java/Miscellaneous/3417-zigzag-grid-traversal-with-skip.java
class Solution {
    public List<Integer> zigzagTraversal(int[][] grid) {
        if (grid == null || grid.length == 0) {
            return null;
        }

        int rows = grid.length;
        int cols = grid[0].length;

        List<Integer> result = new ArrayList<>();

        for (int r = 0; r <= rows - 1; r++) {
            int colStartIndex = 0;
            int colEndIndex = 0;
            int modulo = r%2;
            
            if (modulo == 0) {
                colStartIndex = 0;
                colEndIndex = cols - 1;
                for (int c=colStartIndex; c<=colEndIndex; c=c+2) {
                    int item = grid[r][c];
                    result.add(item);
                }
            } else {

                if (cols % 2 == 0) {
                    colStartIndex = cols - 1;
                } else {
                    colStartIndex = cols - 2;
                }
                colEndIndex = 0;
                for (int c=colStartIndex; c>=colEndIndex; c=c-2) {
                    int item = grid[r][c];
                    result.add(item);
                }
            }
        }

        return result;
    }
}